airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/up_to_date/pipeline.py

```diff
@@ -0,0 +1,326 @@
+            self.logger.info(f"Generated pyproject.toml for {self.context.connector.technical_name}")
+            await with_new_cdk.file("poetry.lock").export(f"{self.context.connector.code_directory}/poetry.lock")
+            self.logger.info(f"Generated poetry.lock for {self.context.connector.technical_name}")
+        except dagger.ExecError as e:
+            return StepResult(step=self, status=StepStatus.FAILURE, stderr=str(e))
+
+        return StepResult(step=self, status=StepStatus.SUCCESS)
+
+
+class RestoreOriginalState(Step):
+    context: ConnectorContext
+
+    title = "Restore original state"
+
+    def __init__(self, context: ConnectorContext) -> None:
+        super().__init__(context)
+        self.pyproject_path = context.connector.code_directory / "pyproject.toml"
+        if self.pyproject_path.exists():
+            self.original_pyproject = self.pyproject_path.read_text()
+        self.poetry_lock_path = context.connector.code_directory / "poetry.lock"
+        if self.poetry_lock_path.exists():
+            self.original_poetry_lock = self.poetry_lock_path.read_text()
+
+    async def _run(self) -> StepResult:
+        if self.original_pyproject:
+            self.pyproject_path.write_text(self.original_pyproject)
+            self.logger.info(f"Restored pyproject.toml for {self.context.connector.technical_name}")
+        if self.original_poetry_lock:
+            self.poetry_lock_path.write_text(self.original_poetry_lock)
+            self.logger.info(f"Restored poetry.lock for {self.context.connector.technical_name}")
+
+        return StepResult(
+            step=self,
+            status=StepStatus.SUCCESS,
+        )
+
+
+class RegressionTest(Step):
+    """Run the regression test for the connector.
+    We test that:
+    - The original dependencies are installed in the new connector image.
+    - The dev dependencies are not installed in the new connector image.
+    - The connector spec command successfully.
+    """
+
+    context: ConnectorContext
+
+    title = "Run regression test"
+
+    async def _run(self, new_connector_container: dagger.Container) -> StepResult:
+        try:
+            await new_connector_container.with_exec(["spec"])
+            await new_connector_container.with_mounted_file(
+                "pyproject.toml", (await self.context.get_connector_dir(include=["pyproject.toml"])).file("pyproject.toml")
+            ).with_exec(["poetry", "run", self.context.connector.technical_name, "spec"], skip_entrypoint=True)
+        except dagger.ExecError as e:
+            return StepResult(
+                step=self,
+                status=StepStatus.FAILURE,
+                stderr=str(e),
+            )
+        return StepResult(
+            step=self,
+            status=StepStatus.SUCCESS,
+        )
+
+
+def compare_semver_versions(version1: str, version2: str) -> int:
```

Comment:
I haven't used it myself but it looks like [packaging.version.Version](https://packaging.pypa.io/en/latest/version.html#packaging.version.Version) does this.

Reply:
I'd suggest using the `semver` package, you can reuse some logic from the [`VersionCheck` step ](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py#L27)

Reply:
random note I learned `semver` crashes with valid PyPi versions like "2.4" and "2.3.1.dev1" - I'm going to use use the python `packaging.version` as it's more likely to work as expected.